primitives/tracing/src/wasm_tracing.rs

```diff
@@ -0,0 +1,2154 @@
+		( $lvl:expr, { $($fields:tt)* }, $($arg:tt)+ ) => { {} };
+		( $lvl:expr, { $($fields:tt)* }, $($arg:tt)+ ) => { {} };
+		($lvl:expr, $($k:ident).+ = $($field:tt)*) => { {} };
+		($lvl:expr, $($k:ident).+, $($field:tt)*) => { {} };
+		($lvl:expr, ?$($k:ident).+, $($field:tt)*) => { {} };
+		($lvl:expr, %$($k:ident).+, $($field:tt)*) => { {} };
+		($lvl:expr, ?$($k:ident).+) => { {} };
+		($lvl:expr, %$($k:ident).+) => { {} };
+		($lvl:expr, $($k:ident).+) => { {} };
+		( $lvl:expr, $($arg:tt)+ ) => { {} };
+	}
+}
+
+#[cfg(feature = "with-tracing")]
+mod inner {
+	use core::{
+		module_path, concat, format_args, file, line,
+	};
+
+	// just a simplistic holder for span and entered spans
+	// that exits on drop
+	pub struct Span(u64);     // 0 means no item
+	pub struct Entered(u64);  // 0 means no item
+
+	impl Span {
+		pub fn new(v: u64) -> Self {
+			Span(v)
+		}
+		pub fn none() -> Self {
+			Span::new(0)
+		}
+		pub fn enter(&self) -> Entered {
+			if self.0 != 0 {
+				crate::with_tracing_subscriber(|t|  t.enter(self.0));
+			}
+			Entered(self.0)
+		}
+		pub fn in_scope<F: FnOnce() -> T, T>(&self, f: F) -> T {
+			let _enter = self.enter();
+			f()
+		}    
+	}
+
+	impl Entered {
+		pub fn exit(&mut self) {
+			if self.0 != 0 {
+				crate::with_tracing_subscriber(|t| t.exit(self.0));
+			}
+		}
+	}
+
+	impl Drop for Entered { 
+		fn drop(&mut self) {
+			self.exit();
+		}
+	}
+
+	/// Constructs a new `Event`.
+	///
+	/// The event macro is invoked with a `Level` and up to 32 key-value fields.
+	/// Optionally, a format string and arguments may follow the fields; this will
+	/// be used to construct an implicit field named "message".
+	///
+	/// See [the top-level documentation][lib] for details on the syntax accepted by
+	/// this macro.
+	///
+	/// [lib]: index.html#using-the-macros
+	///
+	/// # Examples
+	///
+	/// ```rust
+	/// use sp_tracing::{event, Level};
+	///
+	/// # fn main() {
+	/// let data = (42, "forty-two");
+	/// let private_data = "private";
+	/// let error = "a bad error";
+	///
+	/// event!(Level::ERROR, %error, "Received error");
+	/// event!(
+	///     target: "app_events",
+	///     Level::WARN,
+	///     private_data,
+	///     ?data,
+	///     "App warning: {}",
+	///     error
+	/// );
+	/// event!(Level::INFO, the_answer = data.0);
+	/// # }
+	/// ```
+	///
+	// /// Note that *unlike `span!`*, `event!` requires a value for all fields. As
+	// /// events are recorded immediately when the macro is invoked, there is no
+	// /// opportunity for fields to be recorded later. A trailing comma on the final
+	// /// field is valid.
+	// ///
+	// /// For example, the following does not compile:
+	// /// ```rust,compile_fail
+	// /// # #[macro_use]
+	// /// # extern crate sp_tracing;
+	// /// # use sp_tracing::Level;
+	// /// # fn main() {
+	// /// event!(Level::INFO, foo = 5, bad_field, bar = "hello")
+	// /// #}
+	// /// ```
+	#[macro_export]
+	macro_rules! event {
```

Comment:
Why do we need to copy all these macros?

Reply:
These aren't copies but re-implementation with the same syntax. Unfortunately the originals are not supporting `no_std`-environments and use types we can't move across the wasm-rust-boundary (becuase of `'static` lifetime optimizations they have done). And because of the way they are structured we unfortunately can't just inject the most inner actually changed elements but need a copy of all of them. Also not a fan of the bloat but other than cutting down the feature set to below what `tracing` itself offers – like we could just not allow the `trace!` and `trace_event!` shortcut macros – there is not much we can do about it :( .

Reply:
:(